### PR TITLE
change ByteBuffer to extend to all nio.Buffer types

### PR DESCRIPTION
**Pull request by zcaudate**
The main function is `component-size`, which will calculate the size of the nio.Buffer component. It is memoized for speed as the type lookup uses reflection.

### CODE REVIEW

File: src/clojure/uncomplicate/clojurecl/internal/impl.clj [hunk 1]
```diff
@@ -429,12 +429,28 @@
...
+
+(def component-size (memoize component-size-raw))
+
+(extend-type Buffer
   Mem
   (ptr [this]
     (Pointer/toBuffer this))
   (size [this]
-    (.capacity ^ByteBuffer this)))
+    (* (component-size (type this))
```
Comment (zcaudate) [inline src/clojure/uncomplicate/clojurecl/internal/impl.clj]: this is the only real change

Comment (blueberry) [reply src/clojure/uncomplicate/clojurecl/internal/impl.clj]: Did you measure the performance penalty exactly? `size` is used everywhere, so it is very important for it to be as fast as possible.

**Comment by blueberry** [comment]
Hi Chris. Can you please provide tests for the new implementation? The existing tests only cover `ByteBuffer`s, while we have to make sure that this does not cause unintended issues with other kinds of buffers that you've added.

File: src/clojure/uncomplicate/clojurecl/internal/impl.clj (hunk 1 above)
Comment (zcaudate) [reply src/clojure/uncomplicate/clojurecl/internal/impl.clj]: Hmm... memoize is pretty fast, type should be as well, but in that case, the `extend-type` for ByteBuffer should be kept.

**Comment by zcaudate** [comment]
@blueberry: no worries.

File: src/clojure/uncomplicate/clojurecl/internal/impl.clj (hunk 1 above)
Comment (blueberry) [reply src/clojure/uncomplicate/clojurecl/internal/impl.clj]: How many nanoseconds exactly?

**Comment by zcaudate** [comment]
- tests have been added
- ByteBuffer has been added back in

Rough Benchmarks:

```
(comment ;; Benchmarks

  (def -dbuff-  (java.nio.DoubleBuffer/allocate 10))
  (def -tdbuff- (type dbuff))
  (def -trials- 1000000)
  
  (-> (dotimes [i -trials-]
        (component-size-raw (type -dbuff-)))
      time
      with-out-str)
  => "\"Elapsed time: 236.124617 msecs\"\n"
  
  (-> (dotimes [i -trials-]
        (component-size (type -dbuff-)))
      time
      with-out-str)
  => "\"Elapsed time: 215.91447 msecs\"\n"
  
  (-> (dotimes [i -trials-]
        (component-size tdbuff))
      time
      with-out-str)
  => "\"Elapsed time: 181.299622 msecs\"\n"

  (-> (dotimes [i -trials-]
        i)
      time
      with-out-str)
  => "\"Elapsed time: 4.037231 msecs\"\n"

  ;; type speed (ns):
  (- 215.9 181.3)
  => 34.6
  
  ;; memoize speed (ns):
  (- 215.9 4.0)
  => 211.9
  
  ;; non-memoized speed (ns):
  (- 236.12 4.0)
  => 232.12)
```

**Pull request closed without merging**